lms/djangoapps/class_dashboard/dashboard_data.py

```diff
@@ -36,65 +41,120 @@ def get_problem_grade_distribution(course_id):
         attempting the problem
     """
 
-    # Aggregate query on studentmodule table for grade data for all problems in course
-    db_query = models.StudentModule.objects.filter(
-        course_id__exact=course_id,
-        grade__isnull=False,
-        module_type__in=PROB_TYPE_LIST,
-    ).values('module_state_key', 'grade', 'max_grade').annotate(count_grade=Count('grade'))
-
     prob_grade_distrib = {}
-    total_student_count = {}
-
-    # Loop through resultset building data for each problem
-    for row in db_query:
-        curr_problem = course_id.make_usage_key_from_deprecated_string(row['module_state_key'])
-
-        # Build set of grade distributions for each problem that has student responses
-        if curr_problem in prob_grade_distrib:
-            prob_grade_distrib[curr_problem]['grade_distrib'].append((row['grade'], row['count_grade']))
-
-            if (prob_grade_distrib[curr_problem]['max_grade'] != row['max_grade']) and \
-                    (prob_grade_distrib[curr_problem]['max_grade'] < row['max_grade']):
-                prob_grade_distrib[curr_problem]['max_grade'] = row['max_grade']
-
-        else:
-            prob_grade_distrib[curr_problem] = {
-                'max_grade': row['max_grade'],
-                'grade_distrib': [(row['grade'], row['count_grade'])]
-            }
+    total_student_count = defaultdict(int)
+
+    if enrollment <= settings.MAX_ENROLLEES_FOR_METRICS_USING_DB:
+        # Aggregate query on studentmodule table for grade data for all problems in course
+        queryset = models.StudentModule.objects.filter(
+            course_id__exact=course_id,
+            grade__isnull=False,
+            module_type__in=PROB_TYPE_LIST,
+        ).values('module_state_key', 'grade', 'max_grade').annotate(count_grade=Count('grade'))
+
+        # Loop through resultset building data for each problem
+        for row in queryset:
+            curr_problem = course_id.make_usage_key_from_deprecated_string(row['module_state_key'])
+
+            # Build set of grade distributions for each problem that has student responses
+            if curr_problem in prob_grade_distrib:
+                prob_grade_distrib[curr_problem]['grade_distrib'].append((row['grade'], row['count_grade']))
+
+                if (prob_grade_distrib[curr_problem]['max_grade'] != row['max_grade']) and \
```

Comment:
Optional alternative: If you wrap the entire boolean expression in parens, you don't need to use the `\` containuation character.